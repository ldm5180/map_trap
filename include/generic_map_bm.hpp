#ifndef GENERIC_MAP_BM_HPP
#define GENERIC_MAP_BM_HPP

#include "add_to_map.hpp"

#include <boost/hana.hpp>
#include <cstdlib>

static constexpr auto has_reserve =
    boost::hana::is_valid([](auto &&x) -> decltype((void)x.reserve(0)) {});

template <class map_t> static void BM_map_insert(benchmark::State &state) {
  std::srand(42);
  map_t map{};
  if
    constexpr(has_reserve(map)) { map.reserve(state.range(0)); }

  for (auto _ : state) {
    add_to_map(state.range(0), map);
  }
}

template <class map_t> static void BM_map_iterate(benchmark::State &state) {
  std::srand(42);
  map_t map{};
  if
    constexpr(has_reserve(map)) { map.reserve(state.range(0)); }

  for (auto i = 0; i < state.range(0); ++i) {
    add_to_map(state.range(0), map);
  }

  for (auto _ : state) {
    // Nothing to do to sort them. Can simply iterate.
    typename map_t::mapped_type v{};
    for (const auto &kv : map) {
      benchmark::DoNotOptimize(v = kv.second);
    }
  }
}

template <class map_t> static void BM_map_find(benchmark::State &state) {
  std::srand(42);
  map_t map{};
  if
    constexpr(has_reserve(map)) { map.reserve(state.range(0)); }
  std::vector<typename map_t::key_type> keys{};
  keys.reserve(state.range(0));

  for (auto i = 0; i < state.range(0); ++i) {
    add_to_map_and_key_list(state.range(0), map, keys);
  }

  for (auto _ : state) {
    typename map_t::mapped_type v{};
    const auto key = keys[std::rand() % keys.size()];
    benchmark::DoNotOptimize(v = map[key]);
  }
}

#endif // GENERIC_MAP_BM_HPP
