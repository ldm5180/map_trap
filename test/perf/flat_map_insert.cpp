#include "generic_map_bm.hpp"

#include <benchmark/benchmark.h>
#include <boost/container/flat_map.hpp>

BENCHMARK_TEMPLATE(BM_map_insert, boost::container::flat_map<int64_t, int64_t>)
    ->RangeMultiplier(2)
    ->Range(8, 8 << 8);
BENCHMARK_MAIN();
