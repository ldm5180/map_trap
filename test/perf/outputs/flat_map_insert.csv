Run on (4 X 2594.09 MHz CPU s)
2017-10-28 16:39:54
name,iterations,real_time,cpu_time,time_unit,bytes_per_second,items_per_second,label,error_occurred,error_message
"BM_map_insert<boost::container::flat_map<int, std::string>>/8",3000,654147,652753,ns,,,,,
"BM_map_insert<boost::container::flat_map<int, std::string>>/16",1000,880296,880299,ns,,,,,
"BM_map_insert<boost::container::flat_map<int, std::string>>/32",1000,3.40193e+06,3.40125e+06,ns,,,,,
"BM_map_insert<boost::container::flat_map<int, std::string>>/64",503,6.6791e+06,6.67756e+06,ns,,,,,
"BM_map_insert<boost::container::flat_map<int, std::string>>/128",100,5.61039e+06,5.61042e+06,ns,,,,,
"BM_map_insert<boost::container::flat_map<int, std::string>>/256",100,2.23815e+07,2.23809e+07,ns,,,,,
"BM_map_insert<boost::container::flat_map<int, std::string>>/512",74,6.55203e+07,6.55084e+07,ns,,,,,
"BM_map_insert<boost::container::flat_map<int, std::string>>/1024",19,6.46592e+07,6.4657e+07,ns,,,,,
"BM_map_insert<boost::container::flat_map<int, std::string>>/2048",10,1.36916e+08,1.36855e+08,ns,,,,,
