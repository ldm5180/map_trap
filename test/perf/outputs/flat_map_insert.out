Run on (4 X 2594.09 MHz CPU s)
2017-10-28 16:28:18
name,iterations,real_time,cpu_time,time_unit,bytes_per_second,items_per_second,label,error_occurred,error_message
"BM_map_insert<boost::container::flat_map<int, std::string>>/8",3166,653623,653826,ns,,,,,
"BM_map_insert<boost::container::flat_map<int, std::string>>/16",1000,854213,848467,ns,,,,,
"BM_map_insert<boost::container::flat_map<int, std::string>>/32",1000,3.31548e+06,3.30681e+06,ns,,,,,
"BM_map_insert<boost::container::flat_map<int, std::string>>/64",499,6.49717e+06,6.49211e+06,ns,,,,,
"BM_map_insert<boost::container::flat_map<int, std::string>>/128",100,5.3328e+06,5.33535e+06,ns,,,,,
"BM_map_insert<boost::container::flat_map<int, std::string>>/256",100,2.11478e+07,2.11328e+07,ns,,,,,
"BM_map_insert<boost::container::flat_map<int, std::string>>/512",75,6.2592e+07,6.26064e+07,ns,,,,,
"BM_map_insert<boost::container::flat_map<int, std::string>>/1024",20,6.79413e+07,6.7971e+07,ns,,,,,
"BM_map_insert<boost::container::flat_map<int, std::string>>/2048",10,1.33784e+08,1.33849e+08,ns,,,,,
